app/src/main/java/com/amaze/filemanager/ui/fragments/data/MainFragmentViewModel.kt

```diff
@@ -0,0 +1,195 @@
+    var home: String? = null
+    var selection = false
+    var results: Boolean = false
+    lateinit var openMode: OpenMode
+
+    // defines the current visible tab, default either 0 or 1
+    // private int mCurrentTab;
+    /** For caching the back button  */
+    var back: LayoutElementParcelable? = null
+
+    var dragAndDropPreference = 0
+
+    val dataUtils: DataUtils = DataUtils.getInstance()
+    var isEncryptOpen = false // do we have to open a file when service is begin destroyed
+
+    // the cached base file which we're to open, delete it later
+    var encryptBaseFile: HybridFileParcelable? = null
+
+    /** a list of encrypted base files which are supposed to be deleted  */
+    var encryptBaseFiles = ArrayList<HybridFileParcelable>()
+
+    // defines the current visible tab, default either 0 or 1
+    // private int mCurrentTab;
+
+    /*boolean identifying if the search task should be re-run on back press after pressing on
+    any of the search result*/
+    var retainSearchTask = false
+
+    /** boolean to identify if the view is a list or grid  */
+    var isList = true
+    var addHeader = false
+    var accentColor = 0
+    var primaryColor = 0
+    var primaryTwoColor = 0
+    var stopAnims = true
+
+    /**
+     * Initialize arguemnts from bundle in MainFragment
+     */
+    fun initBundleArguments(bundle: Bundle?) {
+        bundle?.run {
+            if (no == 0) {
+                no = getInt("no", 1)
+            }
+
+            if (home.isNullOrBlank()) {
+                getString("home")?.run {
+                    home = this
+                }
+            }
+
+            if (currentPath.isNullOrBlank()) {
+                getString("lastpath")?.run {
+                    currentPath = this
+                }
+            }
+            if (!::openMode.isInitialized) {
+                openMode = if (getInt("openmode", -1) !== -1) {
+                    OpenMode.getOpenMode(getInt("openmode", -1))
+                } else {
+                    OpenMode.FILE
+                }
+            }
+        }
+    }
+
+    /**
+     * Initialize drag drop preference
+     */
+    fun initDragAndDropPreference(sharedPreferences: SharedPreferences) {
+        dragAndDropPreference = sharedPreferences.getInt(
+            PreferencesConstants.PREFERENCE_DRAG_AND_DROP_PREFERENCE,
+            PreferencesConstants.PREFERENCE_DRAG_TO_SELECT
+        )
+    }
+
+    /**
+     * Initialize isList from dataUtils
+     */
+    fun initIsList() {
+        isList = dataUtils.getListOrGridForPath(
+            currentPath,
+            DataUtils.LIST
+        ) == DataUtils.LIST
+    }
+
+    /**
+     * Initialize column number from preference
+     */
+    fun initColumns(sharedPreferences: SharedPreferences) {
+        if (columns == 0) {
+            sharedPreferences.getString(
+                PreferencesConstants
+                    .PREFERENCE_GRID_COLUMNS,
+                "-1"
+            )?.toInt()?.run {
```

Comment:
`sharedPreferences.getString` is never null AFAIK

Reply:
It's marked as Nullable

Reply:
Yes, because you can put `null` in the default. In this case it can't be null AFAIK.

Reply:
But compiler suggest it can be null. Because `getString` returns `String?` and so I have to keep ?. at `toInt` as well. It's all compiler suggested. If I remove I need to use `!!`

Reply:
I thought you could do a soft cast. Nevertheless you can still do a less-than-soft cast:
```kt
val a: String? = sharedPref.getString(PreferencesConstants.PREFERENCE_DIRECTORY_SORT_MODE, "0")
a as String // a will not be null because default value is not null.
```

Reply:
Casting is dangerous. I don't see the urgent need. It's just a null check that too natively supported by kt. In future consider someone changes the default value to null but forgets to remove the casting, will be huge problem then.

Reply:
> Casting is dangerous.

Not here.

> I don't see the urgent need.

Ok.